sample/timestampQuery/TimestampQueryManager.ts

```diff
@@ -0,0 +1,98 @@
+  // or RenderPassDescriptor's 'timestampWrites' field.
+  timestampQuerySet: GPUQuerySet
+
+  // A buffer where to store query results
+  timestampBuffer: GPUBuffer
+
+  // A buffer to map this result back to CPU
+  timestampMapBuffer: GPUBuffer
+
+  // State used to avoid firing concurrent readback of timestamp values
+  hasOngoingTimestampReadback: boolean
+
+  // Device must have the "timestamp-query" feature
+  constructor(device: GPUDevice, timestampCount: number) {
+    this.timestampSupported = device.features.has("timestamp-query");
+    if (!this.timestampSupported) return;
+
+    this.timestampCount = timestampCount;
+
+    // Create timestamp queries
+    this.timestampQuerySet = device.createQuerySet({
+      type: "timestamp",
+      count: timestampCount, // begin and end
+    });
+
+    // Create a buffer where to store the result of GPU queries
+    const timestampByteSize = 8; // timestamps are uint64
+    const timestampBufferSize = timestampCount * timestampByteSize;
+    this.timestampBuffer = device.createBuffer({
+      size: timestampBufferSize,
+      usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.QUERY_RESOLVE,
+    });
+
+    // Create a buffer to map the result back to the CPU
+    this.timestampMapBuffer = device.createBuffer({
+      size: timestampBufferSize,
+      usage: GPUBufferUsage.COPY_DST | GPUBufferUsage.MAP_READ,
+    });
+
+    this.hasOngoingTimestampReadback = false;
+  }
+
+  // Resolve all timestamp queries and copy the result into the map buffer
+  resolveAll(commandEncoder: GPUCommandEncoder) {
+    if (!this.timestampSupported) return;
+
+    // After the end of the measured render pass, we resolve queries into a
+    // dedicated buffer.
+    commandEncoder.resolveQuerySet(
+      this.timestampQuerySet,
+      0 /* firstQuery */,
+      this.timestampCount /* queryCount */,
+      this.timestampBuffer,
+      0, /* destinationOffset */
+    );
+
+    if (!this.hasOngoingTimestampReadback) {
+      // Copy values to the mapped buffer
+      commandEncoder.copyBufferToBuffer(
+        this.timestampBuffer, 0,
+        this.timestampMapBuffer, 0,
+        this.timestampBuffer.size,
+      );
+    }
+  }
+
+  // Once resolved, we can read back the value of timestamps
+  readAsync(onTimestampReadBack: (timestamps: BigUint64Array) => void) {
+    if (!this.timestampSupported) return new Promise(() => {});
+    if (this.hasOngoingTimestampReadback) return new Promise(() => {});
+
+    this.hasOngoingTimestampReadback = true;
+
+    const buffer = this.timestampMapBuffer;
+    return new Promise(resolve => {
+      buffer.mapAsync(GPUMapMode.READ, 0, buffer.size)
+      .then(() => {
+        const rawData = buffer.getMappedRange(0, buffer.size);
+        const timestamps = new BigUint64Array(rawData);
+
+        onTimestampReadBack(timestamps);
+
+        buffer.unmap();
+        this.hasOngoingTimestampReadback = false;
+      })
+    });
+  }
```

Comment:
`resolve` is never called so this promise will never resolve. This function is kind of mixing callback and promise styles, but the returned promise is not used anyway so it could be removed.

```suggestion
  readAsync(onTimestampReadBack: (timestamps: BigUint64Array) => void): void {
    if (!this.timestampSupported) return;
    if (this.hasOngoingTimestampReadback) return;

    this.hasOngoingTimestampReadback = true;

    const buffer = this.timestampMapBuffer;
    void buffer.mapAsync(GPUMapMode.READ)
      .then(() => {
        const rawData = buffer.getMappedRange();
        const timestamps = new BigUint64Array(rawData);

        onTimestampReadBack(timestamps);

        buffer.unmap();
        this.hasOngoingTimestampReadback = false;
      });
  }
```

also nit: the `0, buffer.size` aren't needed because those are the defaults in both mapAsync and getMappedRange

Reply:
Ow my bad, the whole Promise thing is a leftover from a previous attempt where readAsync was returning a Promise **instead** of taking a callback argument, but it was not possible because of mapping lifetime!